### ISSUE TITLE
[Feature Request] provide apis to tell the downstream emitters about the extensible enum info

### ISSUE THREAD

**Issue by qiaozha** [microsoft/typespec#2848]
From https://github.com/Azure/typespec-azure-pr/issues/3827 we are planning to use unions to express extensible enum, but we treat extensible enum different in the emitter compared with regular union. I wonder if compiler could provide some apis to tell the downstream emitter about if we could treat an union as extensible ? As compiler has the information about type relations https://typespec.io/docs/language-basics/type-relations#type-hierarchy 

Another side question is, if we add an unrelated type into the union, which break the extensible enum rule, should we allow it ?
changing `BaseType | SubType1 | SubType2` into `BaseType | SubType1 | SubType2 | UnrelatedType`
as changing from extensible enum to union could be breaking for emitters.  Also, should we add linter rule for that ?

Another reason, I think compiler should provide the detail is that I think the above discussion is for both branding and non-branding scope ?

**Comment by bterlson** [comment]
One API that we should have but don't (as far as I know) is getting a subtype-reduced union, so `BaseType | SubType1 | SubType2` would return just `BaseType`. There are other scenarios where this operation may be useful. And with this I think it would be fairly straight forward to detect the "extensible union" case by seeing if the reduced union has only one variant.

Re: the side question, I don't think TypeSpec itself can error on the addition of an unrelated type - who is to say it's an error? - but I don't have an opinion if Azure wants to lint it. I guess the logic would be, if the subtype reduced union is smaller than the input union but has more than one variant, throw an error?

**Comment by weidongxu-microsoft** [comment]
One case that emitter seems not able to handle is such

```ts
union ColorModel {
  string,
  Red: "Red",
  Blue: "Blue",
  Green: "Green",
}

op getColorModel(): ColorModel;
```

[Playground](https://cadlplayground.z22.web.core.windows.net/cadl-azure/?c=aW1wb3J0ICJAdHlwZXNwZWMvcmVzdCI7Cgp1c2luZyBUeXBlU3BlYy5SZXN0O9AVSHR0cDsKCkBzZXJ2aWNlKHt0aXRsZTogInTERSwgdmVyc2lvbjogIjEuMC4wIn0pCm5hbWVzcGFjZSBUxVIKdW5pb24gQ29sb3JNb2RlbCB7CiAgc3RyaW5nLAogIFJlZDogIlJlZCLEDkJsdcRfxAfFEEdyZWXEX8UIIiwKfQoKb3AgZ2V0ylAoKTrLXjsK&e=%40azure-tools%2Ftypespec-autorest&options=%7B%7D)

typespec-autorest errors
```
error
@azure-tools/typespec-autorest/duplicate-body-types
Request has multiple body types
/test/node_modules/@typespec/compiler/lib/lib.tsp:90:1
```

Compiler gives multiple responses. But the type under `HttpOperationResponse` is just Scalar string. I'd expect it to be that Union.

![image](https://github.com/microsoft/typespec/assets/53292327/d8515f52-fbca-4769-9f9e-0d8a288f6342)

Then there is a change for emitter to interpret it as an extensible enum. Emitter really can't piece together the list of Type that already get broken from the Union.

**Comment by timotheeguerin** [comment]
> One case that emitter seems not able to handle is such
> 
> ```ts
> union ColorModel {
>   string,
>   Red: "Red",
>   Blue: "Blue",
>   Green: "Green",
> }
> 
> op getColorModel(): ColorModel;
> ```
> 
> [Playground](https://cadlplayground.z22.web.core.windows.net/cadl-azure/?c=aW1wb3J0ICJAdHlwZXNwZWMvcmVzdCI7Cgp1c2luZyBUeXBlU3BlYy5SZXN0O9AVSHR0cDsKCkBzZXJ2aWNlKHt0aXRsZTogInTERSwgdmVyc2lvbjogIjEuMC4wIn0pCm5hbWVzcGFjZSBUxVIKdW5pb24gQ29sb3JNb2RlbCB7CiAgc3RyaW5nLAogIFJlZDogIlJlZCLEDkJsdcRfxAfFEEdyZWXEX8UIIiwKfQoKb3AgZ2V0ylAoKTrLXjsK&e=%40azure-tools%2Ftypespec-autorest&options=%7B%7D)
> 
> typespec-autorest errors
> 
> ```
> error
> @azure-tools/typespec-autorest/duplicate-body-types
> Request has multiple body types
> /test/node_modules/@typespec/compiler/lib/lib.tsp:90:1
> ```
> 
> Compiler gives multiple responses. But the type under `HttpOperationResponse` is just Scalar string. I'd expect it to be that Union.
> 
> ![image](https://private-user-images.githubusercontent.com/53292327/305906826-d8515f52-fbca-4769-9f9e-0d8a288f6342.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzNTY3MzIsIm5iZiI6MTcwODM1NjQzMiwicGF0aCI6Ii81MzI5MjMyNy8zMDU5MDY4MjYtZDg1MTVmNTItZmJjYS00NzY5LTlmOWUtMGQ4YTI4OGY2MzQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE5VDE1MjcxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyZjM2MmRlMDVkZjI0N2JlNjE1NmNlZTA0YzQ5ZmUxZDM2YTM2MDg1NzViNTUzZGE3OWQ3ZGZjNDEyYjM0MTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VHN0F1sPflRqlEloDt8sdkUCnNo4phx0NV4hJ1uo_7k)
> 
> Without the ref to the Union, there is almost no chance for emitter to correctly interpret this response as an extensible enum. Emitter really can't piece together the list of Type that already get broken from the Union. (it maybe possible to check back on the `Operation` object, but the parse of its `returnType` likely be too complicated when there are responses in different status code and errors)
That’s is a fair point but I also think this scenario is probably quite the edge case in azure so requiring an explicit @body I think would be fine here.

**Comment by weidongxu-microsoft** [comment]
Thanks Timothee! Yes, the explicit `@body` unblocks the case.

**Comment by qiaozha** [comment]
I want to point out a case. If we have duplicate variants in a union, it's a valid syntax, and according to earlier discussion with @timotheeguerin, typespec won't do type reduce, but if we are expecting this union to express extensible enum, it might be problematic for enum to have duplicate enum members in it.
